### DISCUSSION TITLE
Production server closes websocket after 30 seconds

### DISCUSSION THREAD

**Discussion by lukehutch** [serverpod/serverpod#2528] [Q&A] [answered]
When running on GCE, the server closes the websocket after 30 seconds, every time my app opens it.

However, when running the server locally, the websocket is not closed, it stays open. (I don't understand this discrepancy.)

The only thing I found that might be related to this is 

```dart
    webSocket.pingInterval = const Duration(seconds: 30);
```

in `server.dart`, method `_dispatchWebSocketUpgradeRequest`. However changing this interval doesn't seem to be changing the TTL of the websocket connection.

This will generate more load on the server, because the app has to keep reconnecting. Additionally, when the socket disconnects and reconnects, my app is currently hitting the database to ensure it didn't miss any messages when it was offline. This means every connected client will run a query every 30 seconds, which will increase database load.

Any ideas what may be causing this?

**Comment by lukehutch** [comment]
Sending a ping streaming message from the app to the server after 25 seconds does not prevent the reset after 30 seconds.

**Comment by vlidholt** [comment, marked as answer]
It's not Serverpod that closes the connection, but your load balancer. You may need to configure your load balancer to keep connections open for a (much) longer time or indefinitely, depending on your use case.

**Reply by lukehutch** [reply]
Ah, thanks. What other ramifications of asking the load balancer to never expire websocket connections? I presume the resources are freed if the connection naturally closes?

Shouldn't this be the Terraform script default, to not expire connections?

**Reply by lukehutch** [reply]
@vlidholt I'm looking at `load_balancer.tf`, and I can't figure out where to put a `timeout` parameter. I don't even know where to look in the Terraform and Google docs.

I also have to solve #2522 before I launch, which is probably a load balancer or HTTPS proxy issue too. Would really appreciate your help to get these solved quickly.

**Reply by lukehutch** [reply]
OK, #2522 seems to be resolved now (no idea how I fixed it)...

**Comment by lukehutch** [comment]
OK so for the classic application load balancer:

https://cloud.google.com/load-balancing/docs/https#timeout-bes

> backend service timeout: 30 seconds
> 
> Websocket connections, whether idle or active, automatically close after the backend service times out.
> 
> We don't recommend backend service timeout values greater than 24 hours (86,400 seconds) because Google Cloud periodically restarts GFEs for software updates and other routine maintenance. Your backend service timeout value does not delay the maintenance activities. The longer the backend service timeout value, the more likely it is that Google Cloud terminates TCP connections for maintenance.

I changed the timeout via the Cloud Console, but the change did not fix the problem, and the console tool doesn't show that the timeout actually changed (which, who knows, might be a propagation delay):

```console
$ gcloud compute backend-services describe serverpod-production-backend-web
[...]
timeoutSec: 30
```

I fixed this in the Terraform script:

https://github.com/serverpod/terraform-google-serverpod-cloud-engine/pull/7
